src/htbl.ml

```diff
@@ -0,0 +1,354 @@
+    if not is_found then (false, local)
+    else
+      let curr =
+        match local.curr with
+        | Normal node -> node
+        | _ -> failwith " Should never happen"
+      in
+      if
+        not
+          (Atomic.compare_and_set curr.next local.next
+             (mark_to_be_removed local.next))
+      then unsafe_remove key t
+      else if Atomic.compare_and_set local.prev local.curr local.next then
+        (true, local)
+      else (
+        ignore (find key t);
+        (true, local))
+
+  let remove (key : key) (t : 'a t) = fst (unsafe_remove key t)
+  let mem key t = fst @@ find key t
+end
+
+module Common = struct
+  (* Key value used in the linked list are computed with [reverse] for the
+        dummy node and with [compute_hkey] for the regular node.
+
+     Example of a linked list, simplified to 8 bits integer. Only key
+     are written, as values of regular nodes do not change the linked
+     list order.
+
+      [ 0000 0000 (0, dummy) ]
+     -> [ 0001 0001 (8, regular)]
+     -> [ 0100 0000 (2, dummy)]
+     -> [ 0101 0001 (10, regular)]
+     -> [ 0110 1001 (22, regular)]
+     -> [ 1000 0000 (1, dummy)]
+     -> [ 1000 1001 (17, regular)]
+     -> [ 1100 0000 (3, dummy)]
+  *)
+  let reverse x =
+    (* works for int32 *)
+    let x = x land 0xff_ff_ff_ff in
+    let x = ((x land 0xaa_aa_aa_aa) lsr 1) lor ((x land 0x55_55_55_55) lsl 1) in
+    let x = ((x land 0xcc_cc_cc_cc) lsr 2) lor ((x land 0x33_33_33_33) lsl 2) in
+    let x = ((x land 0xf0_f0_f0_f0) lsr 4) lor ((x land 0x0f_0f_0f_0f) lsl 4) in
+    let x = ((x land 0xff_00_ff_00) lsr 8) lor ((x land 0x00_ff_00_ff) lsl 8) in
+    (x lsr 16) lor (x lsl 16) land 0xffffffff
+
+  let compute_hkey k = reverse k lor 0x00_00_00_01
+
+  (** unset most significant turn on bit (for int32) *)
+  let unset_msb key =
+    let a = key lor (key lsr 1) in
+    let a = a lor (a lsr 2) in
+    let a = a lor (a lsr 4) in
+    let a = a lor (a lsr 8) in
+    let a = a lor (a lsr 16) in
+    (a lsr 1) land key
+
+  let new_dummy_node id llist =
+    let _, local = Llist.unsafe_add (reverse id) Dummy llist in
+    (* If the insertion succeeded ([is_added] = true) then
+       [local.prev] contains the atomic value of the new node.
+
+       If it failed that means another domain has already inserted
+       this dummy node. In this case, [local.prev] also contains the
+       value we seek as the [Llist.find] function calls by
+       [Llist.unsafe_add] will have stopped at the right place. *)
+    Atomic.get local.prev
+end
+
+module Htbl = struct
+  open Common
+  include Type
+
+  type 'a t = { mask : int; buckets : 'a Llist.marked Atomic.t array }
+
+  let init ~size_exponent =
+    let size = Int.shift_left 1 size_exponent in
+    let mask = size - 1 in
+    let llist = Llist.init () in
+    {
+      mask;
+      buckets =
+        Array.init size (fun i ->
+            if i = 0 then Atomic.make (new_dummy_node 0 llist)
+            else Atomic.make Llist.Last);
+    }
```

Comment:
Perhaps we can initialize the buckets on creation in the fixed case? It should improve the performance of hot-path further.